sdk/nodejs/runtime/settings.ts

```diff
@@ -545,3 +549,37 @@ export function monitorSupportsSecrets(): Promise<boolean> {
 export async function monitorSupportsResourceReferences(): Promise<boolean> {
     return monitorSupportsFeature("resourceReferences");
 }
+
+// sxsRandomIdentifier is a module level global that is transfered to process.env.
+// the goal is to detect side by side (sxs) pulumi/pulumi situations for inline programs
+// and fail fast. See https://github.com/pulumi/pulumi/issues/7333 for details.
+const sxsRandomIdentifier = Math.random().toString();
+
+// indicates that the current runtime context is via an inline program via automation api.
+let isInline = false;
+
+/** @internal only used by the internal inline language host implementation */
+export function setInline() {
+    isInline = true;
+}
+
+const pulumiSxSEnv = "PULUMI_NODEJS_SXS_FLAG";
+
+/**
+ * runSxSCheck checks an identifier stored in the environment to detect multiple versions of pulumi.
+ * if we're running in inline mode, it will throw an error to fail fast due to global state collisions that can occur.
+ */
+function runSxSCheck() {
+    const envSxS = process.env[pulumiSxSEnv];
+    process.env[pulumiSxSEnv] = sxsRandomIdentifier;
+
+    if (!isInline) {
+        return;
+    }
+
+    // if we see a different identifier, another version of pulumi has been loaded and we should fail.
+    if (!!envSxS && envSxS !== sxsRandomIdentifier) {
+        throw new Error("Detected multiple versions of '@pulumi/pulumi' in use in an inline automation api program.\n" +
```

Comment:
Would be good to think about how to possibly link to docs with an example resolution in errors like this.